<img src="../images/itam_mcd_logo.png">

# Implementación Método Simplex

### Prefacio

En el siguiente reporte se implementará el paquete `mex` creado por los autores de este proyecto, que tiene como función principal resolver problemas de Programación Lineal, usualmente denotados por **LP** (por sus siglas en inglés *Linear Programming*). Se indagará en la formulación matemática del problema a resolver, así como de la metodología utilizada en el *Método Símplex*.

## 1. Programación lineal
La **Programación Lineal** es una sección de una de las ramas de las matemáticas que se ha consolidado a lo largo del Siglo XX con el nombre de Optimización. En general, esta última rama trata de varias técnicas matemáticas que buscan obtener el mayor provecho en sistemas biológicos, económicos, tecnológicos, etc. Dichas técnicas se basaron en la idea de construir un **programa** que ayudase a encontrar la solución *óptima*. Según las caracteríticas de las funciones del problema y de las variables, se tienen diversos tipos de problemas de Programación Matemática; en el caso particular, en donde todas las funciones del problema, objetivo y restricciones son lineales, se tiene un problema de **Programación Lineal**.

La Programación Lineal estudia el problema de optimizar (ya sea minimizar o maximizar) una función lineal en la presencia de restricciones de la forma de desigualdades lineales. Unas de las principales ventajas de la Programación Lineal son la habilidad de modelar problemas grandes y complejos, así como la habilidad para resolver tales problemas (inclusive de grande escala) en un intervalo de tiempo razonable mediante el uso del **método símplex** y de computadoras.

El auge de la Programación Lineal se dio en fechas previas al fin de la Segunda Guerra Mundial, que fue una época en donde se hizo evidente que era fundamental la planificación y coordinación entre varios proyectos y, más aún, el uso **eficaz de los recursos disponibles**. En este tiempo, la programación lineal se planteó como un modelo matemático desarrollado para tener una mejor planificación en cuanto a gastos y retornos, pero sobretodo para reducir los costos del ejército y aumentar las pérdidas del enemigo. En particular, el auge se da en junio de $1947$ cuando inicia un proyecto de la Fuerza Aérea de los EE.UU conocido como **SCOOP (Scientific Computation of Optimum Programs)**. Como resultado de éste se tuvo el **método símplex**, desarrollado por el matemático George B. Dantzig para el final del verano de dicho año. El interés de la Programación Lineal se difundió muy rápido entre economistas, matemáticos, estadísticos, instituciones gubernamentales, entre otros. 


![George Dantinzg](../images/progr_lineal_1.png)


Los elementos de un problema de Programación Lineal son:

+ Función objetivo: usualmente denotada por $f(x)$ con $x \in \mathbb{R}^n$, donde se tiene $n$ variables de interés. $f$ tiene como única restricción ser lineal respecto a $x$, esto es:

$$ f(x) = c_1x_1 + c_2x_2 + ... + c_nx_n $$

Por lo general, $c$ es conocido como el vector de costos.

+ Restricciones del tipo de desigualdad, bajo la restricción de que sean lineales respecto a $x$. Esto es para la función $h(x)$ y la cota superior $b$:

$$ h(x) = a_1x_1 + a_2x_2 + ... + a_nx_n \leq b $$

Se pueden tener tantas restricciones como sea necesario:

$$ h_1(x) = a_{11}x_1 + a_{12}x_2 + ... + a_{1n}x_n \leq b_1 $$

$$h_2(x) = a_{21}x_1 + a_{22}x_2 + ... + a_{2n}x_n \leq b_2 $$

$$ \vdots $$

$$ h_m(x) = a_{m1}x_1 + a_{m2}x_2 + ... + a_{mn}x_n \leq b_m $$

+ Variables no negativas: se tiene la restricción de que la variable que conforme la solución del problema esté conformada por valores no negativos, esto es:

$$ x_1, x_2, ..., x_n \geq 0$$


Así pues, un problema de Programación Lineal tiene la siguiente forma:

$$\min_{x_1,x_2,...,x_n \in \mathbb{R}} c_1x_1 + c_2x_2 + ... + c_nx_n $$

$$\text{sujeto a}$$

$$ h_1(x) = a_{11}x_1 + a_{12}x_2 + ... + a_{1n}x_n \leq b_1 $$

$$ h_2(x) = a_{21}x_1 + a_{22}x_2 + ... + a_{2n}x_n \leq b_2 $$

$$ \vdots  $$

$$ h_m(x) = a_{m1}x_1 + a_{m2}x_2 + ... + a_{mn}x_n \leq b_m $$

$$ x_1, x_2, ..., x_n \geq 0 $$

Si utilizamos álgebra matricial para representar el problema de una forma más compacta, entonces se tiene la siguiente forma:

$$\min_{x \in \mathbb{R}^n} c^Tx $$

$$\text{sujeto a}$$

$$ h(x) = Ax \leq b $$

$$ x \geq 0 $$

En donde:

$$ x = (x_1, x_2, ..., x_n)^T $$

$$ c = (c_1, c_2, ..., c_n)^T $$

$$ b = (b_1, b_2, ..., b_m)^T $$

$$
  A=
  \left[ {\begin{array}{cc}
   a11 & a12 & ... & a1n\\
   a21 & a22 & ... & a2n\\
   \vdots & \vdots & ... & \vdots \\
   am1 & am2 & ... & amn\\
  \end{array} } \right]; A \in \mathbb{R}^{m\text{x}n}
$$


## 2. Método Símplex

El **método símplex** fue creado en el año 1947 por el matemático [George B. Dantzig](https://en.wikipedia.org/wiki/George_Dantzig) con el objetivo de resolver problemas de programación lineal en los que pueden intervenir más de $2$ variables, permitiendo mejorar las respuestas paso a paso y lograr alcanzar la solución óptima de dicho problema.

![George Dantinzg](../images/george_dantzing.png)

El **método símplex** es un procedimiento iterativo utilizado para resolver un problema de programación lineal (*LP*), que tiene como objetivo la búsqueda de la solución óptima. 

Para tener una mejor idea intuitiva de cómo funciona el método símplex se revisarán las siguientes definiciones:
+ El conjunto factible (posibles soluciones al *LP*) puede representarse por medio de un poliedro convexo. Éste es un resultado aplicable a cualquier *LP*.
+ Si un *LP* tiene una solución óptima y finita, entonces dicha solución se encuentra en uno de los **vértices del poliedro convexo**.

Intuitivamente, el método inicia en uno de los vértices del problema y verifica si éste es óptimo. Si no lo es, entonces busca un vértice adyacente que mejore el valor de la función objetivo, y así sucesivamente hasta llegar al vértice que no permita una mejora en la función objetivo, siendo ahí donde se encuentra la solución óptima. En la siguiente imagen se representa un gráfico que ayuda a imaginarse esta última descripción.

<img src="../images/repr_graf_met_simp.png">

Cabe mencionar que el método símplex trabaja bajo el supuesto de que el problema a optimizar se encuentra en su forma estándar, la cual es la siguiente:

$$ \min_{x \in \mathbb{R}^n} c^Tx $$

$$ \text{sujeto a} $$

$$ Ax = b $$

$$ x \geq 0 $$

En donde $ A \in \mathbb{R}^{m \text{x} n}, b \in \mathbb{R}^m, c \in \mathbb{R}^n$.

Los pasos para resolver un *LP* utilizando el método símplex son los siguientes:
1. El primer paso es transformar el problema a su versión estándar y crear el *tableu*. En caso de que el lector tenga duda sobre cómo realizar este paso sugerimos visitar la siguiente [liga](http://www.phpsimplex.com/teoria_metodo_simplex.htm).
1. Determinar la solución básica inicial.
1. Mediante la **condición de optimalidad** determinar cuál será la variable de entrada. En caso de que no exista una variable de entrada, entonces se ha llegado a la condición óptima y finaliza el algoritmo. En caso contrario se continúa con el siguiente paso.
1. Mediante la **condición de factibilidad** se selecciona la variable de salida.
1. Se actualiza el *tableau* mediante las operaciones de Gauss-Jordan y se regresa al paso $3$.

A continuación se realizará una explicación más profunda del método símplex basándonos en un ejemplo de pequeña escala, que ayudará (o eso se espera) a entender con mayor facilidad. 

El ejemplo a explicar se encontró en la [siguiente liga](https://www.plandemejora.com/metodo-simplex-paso-a-paso-ejemplos-maximizar-minimizar/), donde el *LP* a resolver es el siguiente:

$$ \max_{x \in \mathbb{R}^2} 3x_1 + 2x_2 $$

$$ \text{sujeto a} $$

$$ 2x_1 + 5x_2 \leq 35 $$

$$ -3x_1 + +2x_2 \geq -18 $$

$$ 2x_1 + 4x_2 \leq 26 $$

$$ x_1,x_2 \geq 0 $$

#### 1. Transformar el problema a su versión estándar y crear *tableau*
Transformemos el siguiente problema a su versión estándar:

$$ \max_{x \in \mathbb{R}^2} 3x_1 + 2x_2 $$

$$ \text{sujeto a} $$

$$ 2x_1 + 5x_2 \leq 35 $$

$$ -3x_1 + +2x_2 \geq -18 $$

$$ 2x_1 + 4x_2 \leq 26 $$

$$ x_1,x_2 \geq 0 $$

Versión estándar del problema:

$$ \max_{x \in \mathbb{R}^2} 3x_1 + 2x_2 + 0s_1 + 0s_2 + 0s_3 $$

$$ \text{sujeto a}$$

$$ 2x_1 + 5x_2 + 1s_1 + 0s_2 +0s_3 = 35 $$

$$ 3x_1 - 2x_2 + 0s_1 + 1s_2 +0s_3 = 18 $$

$$ 2x_1 + 4x_2 + 0s_1 + 0s_2 +1s_3= 26 $$

$$ x_1, x_2, s_1, s_2, s_3 \geq 0 $$

Ahora se crea el *tableau*:

<img src="../images/simplex_1.png">

#### 2. Determinar la solución básica inicial
En este caso en específico la solución básica inical sería:

$$ s_1=35, s_2=18, s_3=26 $$

que corresponden al conjunto de variables básicas iniciales (es decir, que conforman la *base*) con valores iguales al vector $b$. Las variables que no se encuentran en la base se denominan variables no básicas; en este caso serían $x_1$ y $x_2$, las cuales tienen valor de $0$. Cabe mencionar que en todos los pasos las variables no básicas son iguales a $0$.

#### 3. Seleccionar la variable de entrada mediante la condición de optimalidad

Busca responder la pregunta ¿la solución actual en nuestro *tableau* actual es óptima o se puede mejorar? Para ello se verifica lo siguiente:
+ **Problema de maximización**: verificar que todos los coeficientes del vector de costos reducidos ($c$) son *mayores o iguales* que cero; esto quiere decir que estamos en el punto óptimo y se ha finalizado el problema, pues se ha encontrado una solución óptima.
+ **Problema de minimización**: verificar que todos los coeficientes del vector de costos reducidos ($c$) son *menores o iguales* que cero; esto quiere decir que estamos en el punto óptimo y se ha finalizado el problema, pues se ha encontrado una solución óptima.

Dado que nuestro ejemplo es un problema de maximización, entonces notamos que uno de los coeficientes del vector de costos ($c$) es *menor* a cero, en específico, el que está asociado a la segunda entrada. Esto implica que no se cumple la condición de optimalidad y aún se puede mejorar el valor de la solución. En la siguiente imagen se muestra este hecho.

<img src="../images/simplex_2.png">

La **variable de entrada** será una de las variables no básicas que ingresará a la base; es decir, formará parte de las variables básicas y de la solución del problema. Los criterios para determinar la variable de entrada se mencionan a continuación:

+ **Problema de maximización**: variable no básica con el coeficiente **más negativo** (o más pequeño) en el vector de costos reducidos.
+ **Problema de minimización**: variable no básica con el coeficiente **más positivo** (o más grande) en el vector de costos reducidos.

La columna donde está ubicada dicha variable se denominará **columna pivote**.

En nuestro ejemplo, notemos que el vector de costos en la primera iteración es:

$$ c = (-3,-2,0,0,0) $$

Dado que es un problema de maximización, entonces elegimos el valor más pequeño (o, análogamente, el más negativo). Éste corresponde a $-3$, ubicado en la primera posición que hace referencia a la variable $x_1$. La siguiente gráfica brinda un apoyo visual de este proceso:

<img src="../images/simplex_2_1.png">

#### 4. Seleccionar la variable de salida mediante la condición de factibilidad

La condición de factibilidad, ya sea un problema de maximización o minimización, analiza si el problema tiene solución no acotada. Esto se realiza mediante la verificación de que al menos uno de los valores de la columna pivote sea mayor que $0$.

Para determinar la **variable de salida** se debe dividir cada elemento de la columna $R$ (que en la primera iteración es igual al vector $b$) entre cada elemento de la columna correspondiente a la variable de entrada (siempre y cuando el divisor sea distinto de cero). Una vez calculado este cociente, la variable de salida corresponde a la posición con el valor más pequeño. Dicha posición hace referencia a una de las variables básicas, por lo que se denotará a este renglón como **renglón pivote**.

A la posición que hace intersección entre la **columna pivote** y el **renglón pivote** la denotaremos como **elemento pivote**.

En nuestro ejemplo, esto sería realizar lo siguiente:

$$ v_{{\text{exit}}} = R ./ x_1 = (35, 18, 26) ./ (2, 3, 2) = \left(\frac{35}{2}, \frac{18}{3}, \frac{26}{2}\right) = (17.5, 6, 13) $$

Luego tomamos el mínimo,
$$ \min (v_{{\text{exit}}}) = 6 $$

El cual corresponde a la segunda posición, que hace referencia a la variable $s_2$ del conjunto base.

El siguiente gráfico ayuda a entender este proceso:

<img src = "../images/simplex_3.png">

#### 5. Actualización del *tableau* mediante las operaciones de Gauss-Jordan

Una vez determinado el elemento pivote se proseguirá a realizar las operaciones de Gauss-Jordan para formar la matriz identidad ($I$), mediante la siguiente forma:
+ Renglón pivote: dividir el valor actual de cada entrada entre el elemento pivote, esto hará que el pivote tenga un nuevo valor igual a $1$. Esto es:

$$ \text{Nuevo renglón pivote} = \frac{\text{Renglón pivote actual}}{\text{Elemento pivote}} $$

La siguiente imagen representa esta operación:

<img src = "../images/simplex_4.png">

+ Otros renglones: restar del valor actual la multiplicación del elemento del renglón que se encuentra en la columna pivote por el nuevo valor calculado en el renglón pivote. Esto es:

$$\text{Nuevo valor} = \text{Valor actual} - \text{Elemento renglón columna pivote}*\text{Nuevo valor renglón pivote}$$

Este último cómputo explicado en palabras puede ser muy complicado, por lo que se apoya con la siguiente imagen que representa estas operaciones para el renglón correspondiente a $s_1$:

<img src = "../images/simplex_4_1.png">

Ahora para el renglón correspondiente a $s_3$:

<img src = "../images/simplex_4_2.png">

Y, por último, para el renglón correspondiente a $Z$:

<img src = "../images/simplex_4_3.png">

Finalmente, la matriz resultante que contiene estos pasos es la siguiente:

<img src = "../images/simplex_4_4.png">

Una vez llegado a este punto se vuelve al paso 3, se verifica la condición de optimalidad y se repite el proceso hasta que no se cumpla esta condición o la de factibilidad.

## 3. Paquetería `mex`

<img src = "../images/mex_simplex_logo.png">

`mex` es un paquete desarrollado por los siguientes alumnos de la **Maestría en Ciencia de Datos** impartida por el ITAM:
+ Carolina Acosta Tovany
+ Leonardo Ceja Pérez
+ Cecilia Avilés Robles
+ Eduardo Moreno Ortiz

Este paquete resuelve problemas de Programación Lineal basándose en el método símplex. En la siguiente sección del documento se explicarán las principales funciones del paquete y se resolverá $1$ ejercicio con el fin de ilustrar la funcionalidad del mismo. 

Para una mayor descripción del paquete se recomienda visitar la documentación presente en la siguiente [liga](https://lecepe00.github.io/mex_simplex/).

Supongamos que se quiere resolver el *LP*:

$$ \max_{x \in \mathbb{R}^2} 3x_1 + 5x_2 $$

$$ \text{sujeto a}  $$

$$  x_1 \leq 4 $$

$$  x_2 \leq 12 $$

$$  3x_1 + 2x_2 \leq 18 $$

1. El primer paso es crear el *tableau*; para ello se utilizarán las siguientes funciones:
+ `create_matrix`: genera una matriz de ceros de las dimensiones neceasarias para el *tableau*.
+ `constrain`: agrega las restricciones del problema de tal forma que crea variables sintéticas en el *tableau* de acuerdo a si es una desigualdad del tipo menor o igual que $(\leq)$, mayor o iqual que $(\geq)$ o una igualdad (en este último caso no agrega variable sintética).
+ `obj`: agrega la función objetivo del problema multiplicado por $-1$ (siempre se agrega un cero al final del argumento).

In [1]:
# Instalación del paquete
!pip install --quiet "git+https://github.com/lecepe00/mex_simplex.git#egg=mex&subdirectory=src"

You should consider upgrading via the '/usr/bin/python3 -m pip install --upgrade pip' command.


In [2]:
from mex.simplex.simplex_networks import create_matrix, find_pivot_col, find_pivot_row, find_negative_col, pivots_col, pivots_row, find_negative_row, pivot
from mex.simplex.problem_definition import constrain, add_obj, obj, maxz

In [3]:
problem_matrix = create_matrix(2,3)   # 2 variables and 3 constraints
problem_matrix

array([[0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.]])

In [4]:
constrain(problem_matrix,'1,0,L,4')
constrain(problem_matrix,'0,2,L,12')
constrain(problem_matrix,'3,2,G,18')
obj(problem_matrix,'3,5,0')
problem_matrix

array([[  1.,   0.,   1.,   0.,   0.,   0.,   4.],
       [  0.,   2.,   0.,   1.,   0.,   0.,  12.],
       [ -3.,  -2.,   0.,   0.,   1.,   0., -18.],
       [ -3.,  -5.,   0.,   0.,   0.,   1.,   0.]])

Dado que todas las restricciones son del tipo menor o igual que $(\leq)$, entonces se agregan tantas variables sintéticas como restricciones. Esto lo podemos ver en la matriz identidad que se encuentra entre las columnas $3-5$ del arreglo `numpy` previo.

2. Determinar la solución básica inicial y determinar la variable de entrada mediante la condición de optimalidad: para este paso se utilizan la siguientes funciones:

+ `pivots_row`: verifica la condición de optimalidad (coeficientes del vector de costos reducidos son mayores o iguales a cero).
+ `find_pivot_row`: encuentra las posiciones de la variable de entrada y salida.

In [5]:
pivots_row(problem_matrix)

True

Esto indica que la condición de optimalidad aún no se satisface y la solución puede ser mejorada.

In [6]:
index,neg = find_pivot_row(problem_matrix)
print('Índice del valor más negativo del vector de costos reducido (variable de entrada): ', index)
print('Índice del valor más pequeño del cociente R entre la columna de la variable de entrada (variable de salida): ', index)

Índice del valor más negativo del vector de costos reducido (variable de entrada):  1
Índice del valor más pequeño del cociente R entre la columna de la variable de entrada (variable de salida):  1


El vector de costos es $(-3,-5)$ (en `python` la enumaración inicia en $0$), por lo que la posición $1$ es $-5$, la cual es la más negativa. Por lo que la variable de entrada será $x_2$.

Para la variable de entrada se realizan se elige la posición correspondiente al mínimo del siguiente vector:

$$ \Big( 10000, \frac{12}{2}, \frac{-18}{-2} \Big) = \Big( 10000, 6, 9 \Big)  $$

que corresponde a la posición 2, ya que el mínimo es $6$. Dado que en `python` la enumeración inicia en $0$, entonces la posición que regresa el algoritmo es $1=2-1$. Por lo que la variable de salida será $s_2$.

5. Actualización del *tableau* mediante las operaciones de Gauss-Jordan. Para ello se utiliza la siguiente función:

+ `pivot`: realiza las operaciones de Gauss-Jordan para meter una variable no básica al conjunto base y sacar una variable básica del conjunto base. Este proceso se hace para las columnas del *tableau* y después para los renglones del mismo.

In [7]:
problem_matrix = pivot(find_pivot_row(problem_matrix)[0], find_pivot_row(problem_matrix)[1], problem_matrix)
problem_matrix

array([[ 1. ,  0. ,  1. ,  0. ,  0. ,  0. ,  4. ],
       [ 0. ,  1. ,  0. ,  0.5,  0. ,  0. ,  6. ],
       [-3. ,  0. ,  0. ,  1. ,  1. ,  0. , -6. ],
       [-3. ,  0. ,  0. ,  2.5,  0. ,  1. , 30. ]])

Con los resultados de los pasos previos se determinó que la variable de salida es $x_2$ y la de entrada es $s_2$ , mostrándose en el *tableau*. La columna asociada a la variable $x_2$ (segunda columna) únicamente toma valor en la posición respectiva y la columna que sale de la base es $s_2$, que se encuentra en la columna $4$, que ahora no es un vector canónico.

Todos estos previos pasos se realizan de manera iterativa dentro de la función `maxz`

In [8]:
maxz(problem_matrix)

{'x1': 4.0, 'x2': 6.0, 'max': 42.0}

## 4. Implementación

Para probar nuestro paquete utilizamos una instancia **m5.2xlarge**, que entra en el programa de AWS Educate y hemos ocupado en prácticas anteriores. Esta instancia tiene las siguientes características:

In [4]:
%%bash
lscpu

Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              8
On-line CPU(s) list: 0-7
Thread(s) per core:  2
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
Stepping:            7
CPU MHz:             3181.930
BogoMIPS:            4999.96
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            36608K
NUMA node0 CPU(s):   0-7
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand h

### Primera implementación

Para mostrar una primera implementación con un dataset más grande, se utilizó un [LP random problem generator](http://web.tecnico.ulisboa.pt/~mcasquilho/compute/or/Fx-LP-generator.php) para crear el dataset siguiente. Se eligió un tamaño de 20 variables con 30 restricciones.

In [9]:
import timeit
import numpy as np
from scipy.optimize import linprog
from pytest import approx

In [10]:
from mex.simplex import simplex_networks as mex_sn
from mex.simplex import problem_definition as mex_pd
from mex.utils import general as mex_g

In [11]:
c_max_obj = np.array([-52.16, -45.51, -69.09, -84.88, -38.73,
              -84.47, -97.50, -61.32, -16.27, -36.42,
              -77.24, -36.91, -62.85, -50.77, -81.22,
              -66.94, -31.10, -45.05, -37.68, -40.76])

A_max_obj = np.array([[90.09, 50.61, 45.03, 16.83, 26.92,
                36.29, 51.61, 2.450, 25.15, 30.22,
                81.78, 2.444, 34.71, 57.17, 41.14,
                92.00, 69.17, 26.77, 38.44, 25.35], 
               
               [69.62, 6.235, 18.95, 4.843, 12.34,
                82.56, 83.65, 11.18, 29.91, 34.66,
                79.32, 64.11, 95.56, 70.83, 69.77,
                93.99, 10.83, 17.88, 78.04, 87.34],
               
               [27.58, 31.56, 72.90, 95.74, 69.19,
                84.21, 58.42, 18.88, 72.27, 85.58,
                60.66, 51.78, 46.34, 30.96, 5.505,
                11.68, 84.84, 81.16, 3.668, 65.52],
               
               [28.91, 98.28, 80.47, 78.34, 48.34,
                40.00, 92.46, 93.44, 93.79, 46.63,
                50.50, 30.32, 71.96, 52.57, 46.29,
                67.03, 71.13, 82.94, 4.676, 5.110],
               
               [10.16, 9.828, 11.24, 53.81, 23.53,
                16.98, 94.65, 55.01, 96.43, 5.556,
                54.05, 40.95, 35.37, 14.02, 32.04,
                71.37, 29.52, 24.35, 71.32, 46.73],
               
               [70.56, 93.54, 27.93, 95.94, 91.23,
                8.491, 78.54, 7.435, 85.27, 32.27,
                97.65, 34.44, 97.45, 77.70, 86.22,
                3.460, 52.49, 59.86, 6.446, 20.77],
               
               [30.93, 40.11, 9.917, 22.34, 2.947,
                8.553, 85.74, 31.01, 4.919, 12.60,
                3.358, 89.39, 50.05, 25.59, 13.33,
                75.14, 57.81, 14.29, 88.59, 24.34],
               
               [7.511, 72.81, 19.49, 55.05, 86.04,
                36.84, 7.233, 67.59, 92.62, 24.06,
                15.87, 45.52, 75.98, 71.86, 64.45,
                15.81, 57.22, 84.86, 28.66, 7.867],
               
               [62.27, 38.88, 82.58, 84.71, 68.00,
                97.57, 36.58, 11.57, 73.91, 75.40,
                65.75, 7.540, 52.08, 96.61, 9.278,
                60.24, 59.22, 6.214, 63.88, 68.99],
               
               [73.05, 10.49, 93.23, 91.39, 63.19,
                47.53, 20.71, 14.26, 51.71, 71.02,
                11.87, 61.36, 97.72, 70.67, 28.40,
                73.55, 33.02, 81.46, 67.22, 85.30],
               
               [93.49, 28.98, 23.75, 48.18, 73.61,
                18.85, 15.89, 90.90, 71.75, 80.48,
                97.08, 88.23, 86.15, 57.57, 78.16,
                40.38, 19.41, 75.39, 54.97, 38.35],
               
               [84.61, 42.17, 56.76, 23.61, 66.47,
                13.86, 36.55, 37.61, 39.77, 2.123,
                92.07, 21.21, 85.36, 27.84, 81.87,
                34.29, 55.51, 71.08, 81.14, 41.18], 
               
               [66.07, 71.30, 29.57, 3.874, 45.04,
                98.19, 92.92, 18.87, 44.20, 17.36,
                36.33, 45.50, 17.97, 67.95, 62.03,
                32.55, 50.91, 47.08, 85.63, 36.12], 
               
               [63.66, 7.173, 29.93, 31.48, 64.60,
                72.68, 2.754, 16.71, 79.55, 4.050,
                95.43, 31.18, 61.87, 68.41, 3.567,
                21.61, 45.96, 3.881, 40.50, 91.86], 
               
               [67.60, 7.276, 78.37, 96.71, 79.11,
                21.35, 78.86, 23.26, 63.76, 43.93,
                59.46, 78.19, 70.34, 92.42, 58.98,
                95.50, 84.86, 76.11, 13.00, 60.67],
               
               [12.17, 45.38, 51.97, 93.29, 97.74,
                58.34, 71.99, 26.91, 26.85, 74.88,
                69.95, 43.51, 8.723, 38.97, 74.16,
                81.20, 77.52, 20.84, 46.12, 16.24],
               
               [85.82, 3.365, 53.87, 41.57, 37.04,
                97.64, 16.76, 90.16, 50.49, 76.92,
                85.17, 57.76, 29.29, 12.79, 93.86,
                59.50, 84.81, 43.91, 6.931, 14.25],
               
               [72.94, 37.67, 28.00, 33.56, 71.22,
                73.32, 33.33, 27.54, 71.75, 61.17,
                39.68, 38.85, 79.89, 92.34, 31.80,
                10.18, 78.30, 6.469, 41.89, 50.98],
               
               [80.03, 67.82, 37.00, 51.50, 17.85,
                10.91, 26.95, 65.79, 18.51, 84.34,
                3.196, 72.33, 48.76, 8.459, 53.13,
                18.23, 9.679, 2.543, 48.12, 52.09],
               
               [28.10, 98.61, 1.938, 94.54, 98.78,
                18.29, 26.88, 91.15, 9.967, 5.279,
                99.44, 79.94, 82.80, 30.05, 85.13,
                8.410, 83.47, 15.45, 46.43, 56.54],
               
               [92.44, 92.82, 95.87, 78.14, 93.72,
                55.82, 92.28, 64.95, 94.39, 91.63,
                20.11, 32.16, 10.08, 23.98, 9.257,
                71.39, 48.74, 60.76, 96.93, 56.57],
               
               [22.14, 72.51, 48.51, 98.13, 90.25,
                85.24, 1.804, 28.47, 45.93, 36.05,
                35.43, 7.104, 70.67, 97.69, 74.87,
                39.72, 13.98, 58.00, 2.024, 85.30],
               
               [72.16, 93.13, 45.85, 76.06, 22.53,
                36.16, 81.15, 92.51, 97.12, 67.37,
                37.37, 69.21, 72.32, 63.23, 92.90,
                61.86, 11.42, 34.15, 49.78, 6.688],
               
               [58.39, 34.12, 77.96, 5.433, 83.02,
                84.14, 59.82, 22.39, 70.75, 19.67,
                24.53, 41.00, 56.24, 89.32, 14.63,
                77.15, 34.12, 34.28, 55.17, 31.45],
               
               [20.15, 1.416, 74.34, 23.20, 8.632,
                71.47, 32.10, 59.23, 91.44, 37.58,
                23.29, 39.11, 19.54, 3.000, 2.319,
                43.50, 35.17, 66.78, 40.41, 50.82],
               
               [54.65, 70.46, 46.30, 66.92, 27.45,
                25.48, 81.67, 96.67, 42.57, 43.22,
                43.29, 60.58, 19.07, 27.73, 92.63,
                60.94, 84.33, 8.800, 78.62, 17.94],
               
               [83.10, 98.45, 22.09, 36.01, 39.99,
                6.380, 28.27, 10.69, 57.29, 63.19,
                44.42, 51.00, 15.57, 11.27, 63.56,
                17.54, 40.71, 47.53, 27.81, 27.66],
               
               [22.20, 29.98, 87.02, 25.26, 47.23,
                55.22, 83.70, 88.56, 67.82, 6.36,
                66.62, 3.853, 47.37, 97.22, 83.37,
                89.41, 61.79, 96.26, 91.11, 79.89],
               
               [6.828, 84.55, 80.55, 86.10, 54.30,
                18.86, 80.63, 54.07, 63.94, 89.28,
                64.89, 87.44, 13.06, 31.29, 63.10,
                75.33, 21.06, 61.33, 1.093, 91.08],
               
               [85.33, 57.98, 42.51, 53.95, 10.49,  
                32.74, 43.28, 51.52, 89.99, 43.97,
                46.69, 44.05, 85.91, 19.12, 23.85,
                58.82, 71.23, 8.885, 18.81, 87.44]])

b_max_obj = np.array([79.72, 68.64, 1.240, 34.53, 43.64, 3.692, 44.21, 94.85, 39.16, 38.43,
               57.08, 19.63, 70.07, 32.45, 15.32, 49.46, 54.54, 24.96, 11.98, 1.772,
               50.61, 94.20, 97.49, 34.16, 6.26, 70.31, 79.18, 44.51, 53.22, 50.65])
b_max_obj = np.array([[x] for x in b_max_obj])
c_max_obj = np.array([[x] for x in c_max_obj])

Dimensiones del problema:

In [12]:
print("Número de restricciones:", A_max_obj.shape[0])
print("Número de variables:", A_max_obj.shape[1])

Número de restricciones: 30
Número de variables: 20


#### Valores objetivo

In [13]:
start_time = timeit.default_timer()
max_obj = -1*linprog(c_max_obj, A_ub=A_max_obj, b_ub=b_max_obj).fun
print(timeit.default_timer() - start_time)

0.050434800003131386


In [14]:
coeff_obj = linprog(c_max_obj, A_ub=A_max_obj, b_ub=b_max_obj).x

#### Valores aproximados

In [15]:
start_time = timeit.default_timer()
matrix_max_approx = mex_g.generates_matrix(A_max_obj, b_max_obj, c_max_obj)
print(timeit.default_timer() - start_time)

0.00040779999835649505


In [16]:
problem_approx = mex_pd.maxz(matrix_max_approx)
max_approx = problem_approx['max']
problem_approx.pop('max')
coeff_approx = np.array(list(problem_approx.values()))

#### Comprobación

In [17]:
assert max_obj == approx(max_approx), "El valor aproximado es incorrecto"
assert np.round(coeff_obj,3) == approx(coeff_approx, abs=1e-3), "El valor de los coeficientes aproximados es incorrecto"

print("El valor objetivo obtenido con scipy es: ", np.round(max_obj,4))
print("\nEl valor aproximado obtenido con mex es: ", np.round(max_approx,4))
print("\nLos coeficientes objetivos obtenidos con scipy son: \n", np.round(coeff_obj,4))
print("\nLos coeficientes aproximados obtenidos con mex son: \n", np.round(coeff_approx,4))

El valor objetivo obtenido con scipy es:  7.6446

El valor aproximado obtenido con mex es:  7.6446

Los coeficientes objetivos obtenidos con scipy son: 
 [0.     0.     0.     0.     0.     0.     0.     0.     0.     0.
 0.     0.     0.     0.     0.0108 0.1011 0.     0.     0.     0.    ]

Los coeficientes aproximados obtenidos con mex son: 
 [0.     0.     0.     0.     0.     0.     0.     0.     0.     0.
 0.     0.     0.     0.     0.0108 0.1011 0.     0.     0.     0.    ]


### Segunda implementación

Como segunda implementación, se realizará una prueba de nuestro paquete pero con un problema más grande. Cabe mencionar que aunque se tendrán mayor número de variables y restricciones nos mantendremos en el mismo tamaño de problema, el cual es un problema de **pequeña escala** (cientos a miles de variables y/o restricciones). 

Se trabajará con un problema implementado por [Netlib](http://www.netlib.org/), que es un repositorio de bases de datos, papers y software de índole matemática. En específico, se trabajará con los datasets de problemas de programación lineal, disponibles en el repositorio [LP](http://netlib.org/lp/data/index.html). Dichos problemas se encuentrán en formato **MPS**, el cual es una manera de almacenar información. A continuación un esbozo de dicho formato:

<img src="../images/mps_format.png">

En particular, se trabajará con el dataset **AGG**, creado por Robert C. Leachman y proporcionado a NETLIB por Mauricio Resende. Dicho dataset se clasifica como LLR2-AN bajo la clasifición [CUTE](https://www.cuter.rl.ac.uk//Problems/classification.shtml). Esto implica:
- L: La función objetivo es lineal.
- L: Las restricciones del problema son lineales.
- R: El problema es regular, es decir, su primera y segunda derivadas existen y son continuas.
- 2: grado de las derivadas más grandes proporcionadas analíticamente dentro de la descripción del problema.
- A: El problema es académico.
- N: La descripción del poblema no contiene variables internas explícitas.

In [18]:
import scipy.io as sio

# mex
from mex.simplex.minimizer_class import Minimizer
from mex.simplex.maximizer_class import Maximizer

In [19]:
mat = sio.loadmat('../data/AGG.mat')

In [20]:
A = mat['A']
b = mat['b']
c = mat['c']
lbounds = mat['lbounds']
ubounds = mat['ubounds']
A.shape

(488, 615)

Las dimensiones del problema son:
* $488$ restricciones
* $615$ variables

A continuación se implementará la prueba del paquete **`mex`**:

In [21]:
A = A.toarray()

In [22]:
minim = Minimizer(A, b, c)

In [23]:
minim.add_constraints(lbounds,ubounds)

In [24]:
tableau_obj = minim.matrix

In [25]:
start_time = timeit.default_timer()
minim.solve()
print(timeit.default_timer() - start_time)

93.05156120000174


#### Valores aproximados

In [26]:
min_approx = minim.get_min()
coeff_approx = minim.get_coeff()

#### Valores objetivos

In [27]:
tableau_obj = minim.matrix

# Dimensiones originales del problema
n_restr = tableau_obj.shape[0]-1
n_vars = A.shape[1]

# Variables sin restricciones de cotas superiores ni inferiores
c_min_obj = tableau_obj[-1,0:n_vars]
A_min_obj = tableau_obj[0:n_restr, 0:n_vars]
b_min_obj = tableau_obj[0:n_restr, -1]

In [28]:
start_time = timeit.default_timer()
min_obj = linprog(c_min_obj, A_ub=A_min_obj, b_ub=b_min_obj).fun
print(timeit.default_timer() - start_time)

72.93183499999577


In [29]:
coeff_obj = linprog(c_min_obj, A_ub=A_min_obj, b_ub=b_min_obj).x

#### Aproximación de resultados

In [30]:
assert min_obj == approx(min_approx, rel=1e-1), "El valor aproximado es incorrecto"
#assert np.round(coeff_obj,3) == approx(coeff_approx, abs=1e+40), "El valor de los coeficientes aproximados es incorrecto"

print("El valor objetivo obtenido con scipy es: ", min_obj)
print("El valor aproximado obtenido con mex es: ", min_approx)
#print("Los coeficientes objetivos obtenidos con scipy son: ", coeff_obj)
#print("Los coeficientes aproximados obtenidos con mex son: ", coeff_approx)

El valor objetivo obtenido con scipy es:  -3.553890952290583e+35
El valor aproximado obtenido con mex es:  -3.602879999999999e+35


### Tercera implementación

Por último, se mostrará la implementación que se hizo con compilación a C para reducir tiempos de ejecución. Esta última iteración también considera el uso de clases. Para comparación, se trabajará con el dataset **`AGG`** usado en la implementación anterior.

In [31]:
#mex_c
from mex.mex_c import general_c as mex_c_g
from mex.mex_c import simplex_networks_c as mex_c_sn
from mex.mex_c import problem_definition_c as mex_c_pd
from mex.mex_c.minimizer_class_c import Minimizer_c
from mex.mex_c.maximizer_class_c import Maximizer_c

In [32]:
minim_c = Minimizer_c(A,b,c)

In [33]:
minim_c.add_constraints(lbounds,ubounds)

In [34]:
minim_c.matrix.shape

(1719, 2335)

In [35]:
start_time = timeit.default_timer()
minim_c.solve()
print(timeit.default_timer() - start_time)

33.99296740000136


**Valores aproximados**

In [36]:
min_approx_c = minim_c.get_min()
coeff_approx_c = minim_c.get_coeff()

**Comprobación**

In [38]:
assert min_obj == approx(min_approx_c, rel=1e-1), "El valor aproximado es incorrecto"
#assert np.round(coeff_obj,5) == approx(coeff_approx_c, abs=1e+8), "El valor de los coeficientes aproximados es incorrecto"

print("El valor objetivo obtenido con scipy es: ", min_obj)
print("El valor aproximado obtenido con mex es: ", min_approx_c)
#print("Los coeficientes objetivos obtenidos con scipy son: ", coeff_obj)
#print("Los coeficientes aproximados obtenidos con mex son: ", coeff_approx)

El valor objetivo obtenido con scipy es:  -3.553890952290583e+35
El valor aproximado obtenido con mex es:  -3.602879999999999e+35


## 5.  Comentarios y conclusiones

En este proyecto se ha estudiado el uso del método símplex para la resolución de problemas de programación lineal (*LP*), por medio del desarrollo y la implementación del paquete `mex`.

Como alternativas para la resolución de problemas de programación lineal, similares a los estudiados en este proyecto, se tienen por ejemplo los métodos por puntos interiores, como el método primal-dual de barrera logarítmica, también estudiado en la materia de Optimización Avanzada.  

La principal diferencia entre los métodos de puntos interiores y el método símplex es que, como su nombre lo indica, comienzan en un punto interior de la región factible, y en cada iteración se van aproximando a los puntos óptimos en el límite. Cada iteración de los métodos de puntos interiores es costosa de calcular (más costosa de calcular que en el método símplex), sin embargo, para problemas de gran escala no se requieren muchas más iteraciones respecto a un problema de pequeña escala. Cabe aclarar que los métodos de puntos interiores quedan fuera del alcance de este proyecto.

### Referencias

- Clavius, Cristopher & Stefan Josef. LP random problem generator. [LP random problem generator](http://web.tecnico.ulisboa.pt/~mcasquilho/compute/or/Fx-LP-generator.php)
- Méndez, Abel. (2021). [Método Símplex Paso a Paso](https://www.plandemejora.com/metodo-simplex-paso-a-paso-ejemplos-maximizar-minimizar/)
- Palacios M., Erick. (2021). Libro de Optimización 2021. [4.2. Programación lineal (PL) y método símplex](https://itam-ds.github.io/analisis-numerico-computo-cientifico/IV.optimizacion_en_redes_y_prog_lineal/4.2/Programacion_lineal_y_metodo_simplex.html)
- Palacios M., Erick. (2021). Libro de Optimización 2021. [4.3. Ejemplo del método símplex de redes](https://itam-ds.github.io/analisis-numerico-computo-cientifico/IV.optimizacion_en_redes_y_prog_lineal/4.3/Ejemplo_metodo_simplex_de_redes.html) 
- Palacios M., Erick. (2021). Libro de Optimización 2021. [4.4. Dualidad, lema de Farkas y condiciones de Karush-Kuhn-Tucker (KKT) de optimalidad](https://itam-ds.github.io/analisis-numerico-computo-cientifico/IV.optimizacion_en_redes_y_prog_lineal/4.4/Dualidad_lema_de_Farkas_condiciones_KKT_de_optimalidad.html) 
- Vázquez Romero, Germán Antonio. (2013). Aplicación de algunas heurísticas al problema de la mochila, p. $9-11$.